src/table/components/virtualized-table/hooks/use-get-hypercube-data-queue.ts

```diff
@@ -0,0 +1,67 @@
+import { useMemo, useRef } from 'react';
+
+export interface AbortablePromise {
+  isCancelled: boolean;
+}
+
+const pageToKey = ({ qLeft, qTop, qWidth, qHeight }: EngineAPI.INxPage) => `${qLeft}-${qTop}-${qWidth}-${qHeight}`;
+
+const useGetHyperCubeDataQueue = (
+  getDataPages: (qPages: EngineAPI.INxPage[]) => Promise<EngineAPI.INxDataPage[]>,
+  handleDataPages: (qDataPages: EngineAPI.INxDataPage[]) => void
+) => {
+  const queued = useRef(new Set<EngineAPI.INxPage>()); // Keep track of all unique pages that should be retrieved
+  const ongoing = useRef(new Set<EngineAPI.INxPage[]>()); // Keep track of ongoing request. Should be aborted if page info or layout is changed
+  const finished = useRef(new Set<string>()); // Keep track of all unqiue pages that have added to the queue
+
+  const queue = useMemo(
+    () => ({
+      enqueue: (page: EngineAPI.INxPage) => {
+        const key = pageToKey(page);
+
+        if (finished.current.has(key)) {
+          return;
+        }
+
+        queued.current.add(page);
+        // Add to finished before it's actually finished. That makes it easier to check if the page should be retrieved or not
+        // as otherwise all three sets (queued, ongoing and finished) have to be queried for presence.
+        finished.current.add(key);
+
+        if (queued.current.size === 1) {
+          queueMicrotask(async () => {
+            const qPages = Array.from(queued.current.values());
+
+            queued.current.clear();
+            ongoing.current.add(qPages);
+
+            try {
+              const qDataPages = await getDataPages(qPages);
+
+              if (ongoing.current.has(qPages)) {
```

Comment:
is this ever false? looks like you add `qPages` then check it right after if it is there?

Reply:
Yes, `ongoing` is accessable via "closure", but that also means that it can be modified from outside the `enqueue()` function.

Hard to explain as it indeed looks like you commented at first glance. But here we go..

If the call to engine in `getDataPages()` takes some time. I simulated this by replacing it with a promise that resolves after 5 seconds. Making the request reaaaaally slow.. The "call stack" could looke something like this:

```js
loadRows() // finished executing:  enqueued data pages (qTop) 0 to 10
queueMicrotask() //  fetching data pages (qTop) 0 to 10
loadRows() // finished executing:  enqueued data pages (qTop) 11 to 20
queueMicrotask() //  fetching data pages (qTop) 11 to 20
// finished executing the queueMicrotask():  called handleDataPages() for data pages (qTop) 0 to 10
// finished executing the queueMicrotask():  called handleDataPages() for data pages (qTop) 11 to 20
```

But in a scenario when the queue is cleared by calling `queue.clear()`, this happens from `useInfiniteScrollData` when `layout` or `pageInfo` deps changes. The logged "call stack" would look more like this:
```js
loadRows() // finished executing:  enqueued data pages (qTop) 0 to 10
queueMicrotask() //  fetching data pages (qTop) 0 to 10
loadRows() // finished executing:  enqueued data pages (qTop) 11 to 20
queueMicrotask() //  fetching data pages (qTop) 11 to 20
queue.clear(); // was called from "useInfiniteScrollData" because layout or pageInfo changed
// "ongoing.current" set is now empty. So "ongoing.current.has(qPages)" is false.
// handleDataPages() is never called
```